contracts/AssetPool.sol

```diff
@@ -1,41 +1,56 @@
 // SPDX-License-Identifier: MIT
-pragma solidity >=0.4.22 <0.9.0;
+pragma solidity ^0.8.0;
 import "./LiquidityPoolInterfaces.sol";
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+import "hardhat/console.sol";
+import "./Proxy.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
 
-contract AssetPool is LiquidityPoolInterfaces, Ownable, ReentrancyGuard, ERC20("Cruize Protection LP Token", "cr") {
+contract AssetPool is Proxy, Ownable, ReentrancyGuard {
   using SafeMath for uint256;
-  mapping(address => mapping(address => uint)) public userInfo;
+  mapping(address => mapping(address => uint256)) public userInfo;
   mapping(address => bool) private allowedAssets;
+  mapping(address => address) public assetToLp;
+  address[] public tokenAddress;
+
+
+  event Withdraw(
+      address indexed account,
+      uint256 writeAmount
+  );
+  event Provide(address indexed account,uint256 writeAmount);
+
+  // This is proxy code
+  address immutable masterContract;
+
+  constructor(address _masterContract) {
+    masterContract = _masterContract;
+  }
+
+  function createToken(string memory name_, string memory symbol_, address assetAddress) external onlyOwner nonReentrant {
+    CruizeToken token = CruizeToken(createClone(masterContract));
+    token.init(name_, symbol_);
+    assetToLp[assetAddress] = address(token);
+  }
+  // Proxy code ends here
 
   function depositAsset(uint256 depositAmount, address assetAddress) external payable nonReentrant {
     require(depositAmount > 0, "1: Deposit Amount cannot be 0");
-    require(allowedAssets[assetAddress] == true, "Asset not approved for deposit");
-    userInfo[msg.sender][assetAddress] += (depositAmount);
-    IERC20 depositToken = IERC20(assetAddress);
+    userInfo[msg.sender][assetAddress] += depositAmount;
+    ILPtoken depositToken = ILPtoken(assetAddress);
     require(depositToken.transferFrom(msg.sender, address(this), depositAmount), "1: Transfer failed!");
-    _mint(msg.sender, depositAmount);
+    ILPtoken(assetToLp[assetAddress]).mint(msg.sender,depositAmount);
     emit Provide(msg.sender, depositAmount);
   }
 
-  function withdrawAsset(uint crAmount, address assetAddress) external payable nonReentrant {
-    require(crAmount <= balanceOf(msg.sender), "1: Amount is too large");
-    require(crAmount > 0, "1: Amount cannot be zero");
-    require(allowedAssets[assetAddress] == true, "Asset not approved for withdrawal");
+  function withdrawAsset(uint256 assetsValue, address assetAddress) external payable nonReentrant {
```

Comment:
the assetvalue can change to amount and the asset address could be changed with the token.